## Importar librerías

In [10]:
import pandas as pd
import numpy as np

## Leer Dataset

*1. Leer Dataset con pandas (tip: pd.read_csv).*

In [36]:
#Ruta donde se va a leer el dataset
path= './../../dataset/demanda-de-electricidad-datos-mensuales.csv'
#Se le asigna a una variable de python el dataset por medio de la funcion de Pandas "pd.read_csv()"
#dataset = pd.read_csv(path, parse_dates=['indice_tiempo']) 
dataset = pd.read_csv(path)

### Primera exploración del dataset con pandas.
###### .head()
    la funcion head te muestra las 5 primeras filas de tu dataset.

In [3]:
dataset.head(3)

indice_tiempo  demanda_total  demanda_residencial  comercio_e_industria  \
0    2001-01-01        6866.26                  NaN                   NaN   
1    2001-02-01        6473.10                  NaN                   NaN   
2    2001-03-01        6952.30                  NaN                   NaN   

   grandes_usuarios  temperatura_promedio  potencia_maxima  ede_tucuman  \
0               NaN                25.432          13501.0          NaN   
1               NaN                25.596          14061.0          NaN   
2               NaN                22.852          13780.0          NaN   

   edelap_sa  edenor_distribuidor  ...  energia_de_catamarca_sa  \
0        NaN                  NaN  ...                      NaN   
1        NaN                  NaN  ...                      NaN   
2        NaN                  NaN  ...                      NaN   

   energia_de_entre_rios_sa  energia_de_mendoza_sa  \
0                       NaN                    NaN   
1                       NaN                    NaN   
2                       NaN                    NaN   

   energia_san_juan_sa_exedessa  epec_distribuidor  epen_distribuidor  \
0                           NaN                NaN                NaN   
1                           NaN                NaN                NaN   
2                           NaN                NaN                NaN   

   epesf_distribuidor  recursos_y_energia_formosa_sa  secheep  spse_santa_cruz  
0                 NaN                            NaN      NaN              NaN  
1                 NaN                            NaN      NaN              NaN  
2                 NaN                            NaN      NaN              NaN  

[3 rows x 30 columns]

## Columnas del Dataset

*2. ¿Cuantas columnas tiene el dataset? ¿ qué nombres tienen?.  * <br>
*3. ¿Qué tipos de datos tiene las columnas? *

In [14]:
len(dataset.columns)

30

In [20]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 30 columns):
indice_tiempo                     214 non-null datetime64[ns]
demanda_total                     214 non-null float64
demanda_residencial               166 non-null float64
comercio_e_industria              166 non-null float64
grandes_usuarios                  166 non-null float64
temperatura_promedio              214 non-null float64
potencia_maxima                   214 non-null float64
ede_tucuman                       58 non-null float64
edelap_sa                         58 non-null float64
edenor_distribuidor               58 non-null float64
edesal_distribuidor               58 non-null float64
edestesa_emp_dist__este           58 non-null float64
edesur_distribuidor               58 non-null float64
emp_de_energia_de_rio_negro_sa    58 non-null float64
emp_dist_energ_atlantica          58 non-null float64
emp_de_energia_de_la_rioja_sa     58 non-null float64
emp_dist_energia_

##### ¿Cuantas columnas tiene el data set?

In [13]:
dataset.shape

(214, 30)

##### ¿Cuales son las dimensiones del Dataset?

In [5]:
dataset.shape

(214, 30)

   Tenemos **213 Filas** y **30 Columnas**

## Tipos de columnas
#### .dtypes 
    Es un atributo del DataFrame que te da los tipos de datos por cada Columna

In [21]:
dataset.dtypes

indice_tiempo                     datetime64[ns]
demanda_total                            float64
demanda_residencial                      float64
comercio_e_industria                     float64
grandes_usuarios                         float64
temperatura_promedio                     float64
potencia_maxima                          float64
ede_tucuman                              float64
edelap_sa                                float64
edenor_distribuidor                      float64
edesal_distribuidor                      float64
edestesa_emp_dist__este                  float64
edesur_distribuidor                      float64
emp_de_energia_de_rio_negro_sa           float64
emp_dist_energ_atlantica                 float64
emp_de_energia_de_la_rioja_sa            float64
emp_dist_energia_de_salta                float64
emp_electric_misiones_sa                 float64
empresa_dis_estero_sa                    float64
empresa_jujenia_de_energia_sa            float64
energia_de_catamarca

#### Que tipos de datos tiene Pandas
* **object**: Texto
* **float**: Numerico con decimales
* **int**: Numerico entero
* **bool**: True o False, Valores binarios.
* **category**: Valores categoricos
* **datetime ó timedelta**: Son Datos con fechas o tiempos

## Subset de los datos generales y la compañía Epec

* 4.Obtener sólo las columnas ['indice_tiempo', 'demanda_residencial', 'comercio_e_industria', 'grandes_usuarios', 'temperatura_promedio', 'epec_distribuidor'] *

In [28]:
subset = dataset[['indice_tiempo','demanda_residencial','comercio_e_industria', 'grandes_usuarios', 'temperatura_promedio', 'epec_distribuidor']]
subset.head(2)

indice_tiempo  demanda_residencial  comercio_e_industria  grandes_usuarios  \
0    2001-01-01                  NaN                   NaN               NaN   
1    2001-02-01                  NaN                   NaN               NaN   

   temperatura_promedio  epec_distribuidor  
0                25.432                NaN  
1                25.596                NaN

### Podemos buscar solo los datos que sean del 2017 y 2018
###### Con la funcion de Pandas .loc

*5. A partir de “indice_tiempo”, crear Columna  “Anio” (Tips str.slice() )*

In [38]:
subset['Anio'] = dataset['indice_tiempo'].str.slice(stop=4) #probando = prob (0,1,2,3) 
subset.head(2)

indice_tiempo  demanda_residencial  comercio_e_industria  grandes_usuarios  \
0    2001-01-01                  NaN                   NaN               NaN   
1    2001-02-01                  NaN                   NaN               NaN   

   temperatura_promedio  epec_distribuidor  Anio  
0                25.432                NaN  2001  
1                25.596                NaN  2001

*7. Mostrar sólo los datos de 2017 y 2018 *

In [41]:
filter_1 = subset['Anio'].isin(['2017', '2018'])

filter_1.head(2)

0    False
1    False
Name: Anio, dtype: bool

In [52]:
subset2 = subset[filter_1]
subset2.head(4)

indice_tiempo  demanda_residencial  comercio_e_industria  \
192    2017-01-01          5681.816626           4849.049020   
193    2017-02-01          4618.226100           4551.680399   
194    2017-03-01          4470.715122           4514.516433   
195    2017-04-01          3754.330119           3986.428466   

     grandes_usuarios  temperatura_promedio  epec_distribuidor  Anio  
192            1994.0                  25.7         864.972105  2017  
193            1905.0                  25.4         750.517639  2017  
194            2137.0                  22.4         769.956276  2017  
195            2053.0                  18.9         671.747233  2017

##### Cuál es el promedio de la demanda total para el 2018?

*8. Obtener el promedio de la demanda_residencial (Tip:  mean() )*

In [56]:
promedio = subset2.demanda_residencial.mean()

print("{:.2f}".format(promedio))

4734.14


**Contar los datos null en cada columna**

*9. Contar los datos null en cada columna (Tip: isnull() + sum()*

In [57]:
subset.isnull().sum()

indice_tiempo             0
demanda_residencial      48
comercio_e_industria     48
grandes_usuarios         48
temperatura_promedio      0
epec_distribuidor       156
Anio                      0
dtype: int64

*10. Completar los datos nulos de demanda_residencial con el promedio *

In [66]:
subset.demanda_residencial.fillna(subset.demanda_residencial.mean(), inplace=True)

subset.demanda_residencial.isnull().values.any()

False

*11. Por año, cuáles fueron los 3 meses que mas demanda_residencial hubo luego del 2015? *

In [70]:
subset.demanda_residencial.sort_values(ascending=False).head(3)

210    6166.950992
192    5681.816626
180    5651.444906
Name: demanda_residencial, dtype: float64

In [77]:
subset.loc[subset.Anio > '2015'].sort_values('demanda_residencial', ascending=False).groupby("Anio")['indice_tiempo', 'demanda_residencial'].head(2)

indice_tiempo  demanda_residencial
210    2018-07-01          6166.950992
192    2017-01-01          5681.816626
180    2016-01-01          5651.444906
185    2016-06-01          5634.190447
209    2018-06-01          5587.021997
198    2017-07-01          5460.615925

¿cuál es la diferencia entre data set y data frame? ¿Qué significan los dos corchetes? ¿Por qué no toma el indice tiempo como datetime?  
Como se utiliza el slice, para crear una columna "anio" dentro de otra columna "indice_tiempo.


In [78]:
list = [1,2,3,4,5,6,7,8]

len(list)


8

In [81]:
list = ['indice_tiempo','demanda_residencial','comercio_e_industria', 'grandes_usuarios', 'temperatura_promedio', 'epec_distribuidor']

subset3 = dataset[list]

subset3.head(4)

indice_tiempo  demanda_residencial  comercio_e_industria  grandes_usuarios  \
0    2001-01-01                  NaN                   NaN               NaN   
1    2001-02-01                  NaN                   NaN               NaN   
2    2001-03-01                  NaN                   NaN               NaN   
3    2001-04-01                  NaN                   NaN               NaN   

   temperatura_promedio  epec_distribuidor  
0                25.432                NaN  
1                25.596                NaN  
2                22.852                NaN  
3                18.317                NaN

In [ ]:
loc para seleccionar filas y columnas al mismo tiempo